build x86_64-centos7-linux-gnu error

I use sample config centos7-linux-gnu build from ubuntu 24.04
```
[CFG  ]      checking for ld --no-whole-archive... no
[ERROR]      configure: error: support for --no-whole-archive is needed
[ERROR]  
[ERROR]  >>
[ERROR]  >>  Build failed in step 'Building for multilib 1/1: '''
[ERROR]  >>        called in step 'Installing C library'
[ERROR]  >>        called in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@378]
[ERROR]  >>        called from: glibc_backend_once[scripts/build/libc/glibc.sh@209]
[ERROR]  >>        called from: CT_IterateMultilibs[scripts/functions@1610]
[ERROR]  >>        called from: glibc_main[scripts/build/libc/glibc.sh@26]
[ERROR]  >>        called from: do_libc_main[scripts/build/libc.sh@33]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@710]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: 'build.log'
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      https://crosstool-ng.github.io/docs/known-issues/
[ERROR]  >>
[ERROR]  >>  If you feel this is a bug in crosstool-NG, report it at:
[ERROR]  >>      https://github.com/crosstool-ng/crosstool-ng/issues/
[ERROR]  >>
[ERROR]  >>  Make sure your report includes all the information pertinent to this issue.
[ERROR]  >>  Read the bug reporting guidelines here:
[ERROR]  >>      http://crosstool-ng.github.io/support/
[ERROR]  
[ERROR]  (elapsed: 33:18.21)
```

I have checked --no-whole-archive is suppored.
```bash
./x86_64-centos7-linux-gnu-ld --help | grep -- '--no-whole-archive'
  --no-whole-archive          Turn off --whole-archive
```



- Report crosstool-NG version
build from latest master 
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
- Report host OS and its version
Ubuntu 24.04.1 LTS (GNU/Linux 6.8.0-45-generic x86_64)
- Attach `.config` (unless the issue is not related to any configuration)
[config.txt](https://github.com/user-attachments/files/17173507/config.txt)
- Attach `build.log` (unless crosstool-NG fails before the build starts)
[build.log](https://github.com/user-attachments/files/17173496/build.log)

For posterity the problem is the configure check in glibc
```
configure:6911: checking for ld --no-whole-archive
configure:6924: x86_64-centos7-linux-gnu-gcc  -g -O2 -fcommon -U_FORTIFY_SOURCE          -Wno-missing-attributes -Wno-array-bounds -Wno-array-parameter -Wno-stringop-overflow -Wno-maybe-uninitialized    
			    -nostdlib -nostartfiles -Wl,--no-whole-archive
			    -o conftest conftest.c 1>&5
conftest.c:1:1: error: return type defaults to 'int' [-Wimplicit-int]
    1 | _start () {}
      | ^~~~~~
conftest.c:3:1: error: return type defaults to 'int' [-Wimplicit-int]
    3 | __throw () {}
      | ^~~~~~~
configure:6927: $? = 1
configure:6935: result: no
configure:6938: error: support for --no-whole-archive is needed
```

GCC 14 has started treating lazy code like `foo() {}` as an error because it really should be `void foo(void) {}`. As @QBos07 already implied this is fixed in newer glibc versions so updating GLIBC or using an older GCC will avoid the issue.

I'm still wondering if we should throw in something like `CFLAGS=-Wno-error=implicit-int` on the ct-ng side to avoid such issues.

> For posterity the problem is the configure check in glibc
> 
> ```
> configure:6911: checking for ld --no-whole-archive
> configure:6924: x86_64-centos7-linux-gnu-gcc  -g -O2 -fcommon -U_FORTIFY_SOURCE          -Wno-missing-attributes -Wno-array-bounds -Wno-array-parameter -Wno-stringop-overflow -Wno-maybe-uninitialized    
> 			    -nostdlib -nostartfiles -Wl,--no-whole-archive
> 			    -o conftest conftest.c 1>&5
> conftest.c:1:1: error: return type defaults to 'int' [-Wimplicit-int]
>     1 | _start () {}
>       | ^~~~~~
> conftest.c:3:1: error: return type defaults to 'int' [-Wimplicit-int]
>     3 | __throw () {}
>       | ^~~~~~~
> configure:6927: $? = 1
> configure:6935: result: no
> configure:6938: error: support for --no-whole-archive is needed
> ```
> 
> GCC 14 has started treating lazy code like `foo() {}` as an error because it really should be `void foo(void) {}`. As @QBos07 already implied this is fixed in newer glibc versions so updating GLIBC or using an older GCC will avoid the issue.
> 
> I'm still wondering if we should throw in something like `CFLAGS=-Wno-error=implicit-int` on the ct-ng side to avoid such issues.

I added `-Wno-implicit-int` in config,  but still not working.  Even I tried to use gcc 13.3, same error message was shown.

```[EXTRA]      Configuring C library
[EXTRA]      Building C library
[ERROR]      make[3]: *** Cannot open jobserver /tmp/GMfifo3569663r: No such file or directory.  Stop.
[ERROR]      make[2]: *** [Makefile:233: csu/subdir_lib] Error 2
[ERROR]      make[1]: *** [Makefile:9: all] Error 2
[ERROR]   
[ERROR]  >>
[ERROR]  >>  Build failed in step 'Building for multilib 1/1: '''
[ERROR]  >>        called in step 'Installing C library'
[ERROR]  >>        called in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@378]
[ERROR]  >>        called from: glibc_backend_once[scripts/build/libc/glibc.sh@268]
[ERROR]  >>        called from: CT_IterateMultilibs[scripts/functions@1610]
[ERROR]  >>        called from: glibc_main[scripts/build/libc/glibc.sh@26]
[ERROR]  >>        called from: do_libc_main[scripts/build/libc.sh@33]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@712]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: 'build.log'
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      https://crosstool-ng.github.io/docs/known-issues/
[ERROR]  >>
[ERROR]  >>  If you feel this is a bug in crosstool-NG, report it at:
[ERROR]  >>      https://github.com/crosstool-ng/crosstool-ng/issues/
[ERROR]  >>
[ERROR]  >>  Make sure your report includes all the information pertinent to this issue.
[ERROR]  >>  Read the bug reporting guidelines here:
[ERROR]  >>      http://crosstool-ng.github.io/support/
[ERROR]   
[ERROR]  (elapsed: 9:11.60)
[09:13] / gmake: *** [/usr/local/bin/ct-ng:261: build] Error 2
```
[config.txt](https://github.com/user-attachments/files/17321814/config.txt)
[build.log](https://github.com/user-attachments/files/17321816/build.log)

--- a/config/libc/glibc.in
+++ b/config/libc/glibc.in
@@ -197,8 +197,8 @@ config GLIBC_ENABLE_DEBUG
 config GLIBC_EXTRA_CFLAGS
     string
     prompt "extra target CFLAGS"
-    default "-Wno-missing-attributes -Wno-array-bounds -Wno-array-parameter -Wno-stringop-overflow -Wno-maybe-uninitialized" if GLIBC_2_29_or_older && GCC_11_or_later
-    default ""
+    default "-Wno-missing-attributes -Wno-array-bounds -Wno-array-parameter -Wno-stringop-overflow -Wno-maybe-uninitialized" if GLIBC_2_29_or_older && GCC_11_or_later && !GCC_14_or_later
+    default "-Wno-missing-attributes -Wno-array-bounds -Wno-array-parameter -Wno-stringop-overflow -Wno-maybe-uninitialized -Wno-implicit-int" if GLIBC_2_29_or_older && GCC_14_or_later
     help
       Extra target CFLAGS to use when building.
 

